### PR TITLE
fix(agent): skill visibility, selection enforcement, and generic model building (#162)

### LINKED ISSUE #162: feat(agent): audit and enhance agent tools — memory, code editing, file operations
## Context

After the LangGraph migration, the agent has 12 tools but several need review and enhancement. The current tool set lacks some capabilities that modern agent systems (Claude Code, OpenClaw, SWE-agent) provide as standard.

### 1. Memory tool (`update_session_config`) — purpose unclear
- Current tool only sets `analysisType`, `designCode`, `skillIds` — essentially session config
- No persistent memory across conversations (no way to remember user preferences, project context, past decisions)
- Need to evaluate: should this be renamed? Should a separate memory/knowledge tool be added?

### 2. No code self-modification capability
- Agent cannot modify its own skill definitions, prompts, or tool implementations
- Claude Code and similar agents can read/edit workspace files including source code
- Consider: should the agent be able to edit Python runtime scripts, skill YAML, or handler.ts files?

### 3. File tools are basic
- `read_workspace_file` / `write_workspace_file` / `list_workspace_files` work but are minimal
- Missing: file search by content (grep), file diff/patch, file rename/move/delete
- Reference implementations: Claude Code has separate Glob (file discovery) + Grep (content search with 3 output modes + pagination)

### 4. No terminal/command execution tool
- OpenClaw and Claude Code can run shell commands in a sandboxed environment
- Useful for: installing Python packages, running validation scripts, debugging analysis failures

## Proposed approach

Audit all 12 existing tools, then design enhancements inspired by:
- **Claude Code**: Glob + Grep + Read + Write + Bash (separate concerns, composable)
- **OpenClaw**: file operations + command execution + code search
- **SWE-agent**: `find_file` + `search_dir` + `search_file` + `edit_file`

### Priority tools to add/fix
1. **Content search tool** (grep in workspace) — high value for debugging
2. **Memory/knowledge tool** — persist user context across conversations
3. **Command execution tool** (sandboxed) — enable package install, script debug
4. **Rename `update_session_config`** — clarify its purpose or split into session config + memory

## Related

- PR #156 (LangGraph migration)
- Current tool definitions: `backend/src/agent-langgraph/tools.ts`

### PR DESCRIPTION
**Pull request by qinsz01**
## Summary

- **Frontend skill visibility**: Skills not displaying because API returns `{skills:[...]}` wrapper but frontend expected bare array. Fixed parsing in `ai-console.tsx` and `capability-settings-panel.tsx`.
- **Skill selection enforcement**: Added `skillScope` to `AgentConfigurable`, set once in `graph.ts` `toolsNode`. All 5 engineering tools read skill scope from this single point — LLM cannot bypass user selection via `skillIdsJson` or `engineId` parameters (removed from schemas).
- **Generic model building**: `tryBuildGenericModelWithLlm` call path was dead code after LangGraph refactor (#156). Restored as fallback in `skillRuntime.buildModel()` when `handler.buildModel()` returns undefined for generic plugin.
- **Streaming skillId**: Added `extractSkillIdFromContent()` to `streaming.ts` and skillId badge to `ToolCallCard` for frontend visibility.
- **Redundant code cleanup**: Removed duplicate `ToolDeps`/`createAllTools` wrapper, eliminated circular import `tools.ts ↔ tool-registry.ts`.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [x] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [ ] Gateway / orchestration
- [x] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [ ] Integrations
- [x] API / contracts
- [x] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #162
- Related #156 (LangGraph refactor that introduced the regression)

## Root Cause / Regression History (if applicable)

- Root cause: LangGraph refactor (#156) dropped the `tryBuildGenericModelWithLlm` fallback from `skillRuntime.buildModel()`, making generic skill model building impossible. Additionally, `skillIdsJson`/`engineId` in tool schemas let LLM bypass user skill selection.
- Missing detection / guardrail: No enforcement that only user-selected skills are visible to tools. Tool schemas accepted skill/engine IDs as LLM-provided parameters.
- Prior context: PR #156 replaced the deterministic pipeline with ReAct agent. The old `buildModelFromDraft` had a two-layer fallback (deterministic → LLM) that was lost in the rewrite.
- Why this regressed now: The generic skill handler intentionally returns `undefined` from `buildModel()` (no deterministic builder), relying on the LLM fallback that was deleted.

## Regression Test Plan (if applicable)

1. All 641 backend tests pass (`npm test --prefix backend -- --runInBand`)
2. Frontend build and type-check pass (`npm run build --prefix frontend`, `npm run type-check --prefix frontend`)
3. Manual test: "设计一个简支梁，跨度10m，梁中间荷载1kN" — generic skill should detect type, extract params, build model, run analysis
4. Verify skillId badge appears in ToolCallCard during execution

## Evidence

- [x] Failing test/log before + passing after
- [ ] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

All 641 backend tests pass. Frontend build + type-check pass.

## Human Verification (required)

- Verified scenarios: Backend build + 641 tests pass. Frontend build + type-check pass.
- Edge cases checked: Verified `set_session_config` still accepts `skillIdsJson` (legitimate write-to-state use). Verified `run_code_check`/`validate_model` `engineId` untouched (different concern, not skill-selection bypass).
- What you did **not** verify: End-to-end LLM integration test (requires `LLM_API_KEY`).

## Compatibility / Migration

- Backward compatible? (`Yes`)
- Config/env changes? (`No`)
- Migration needed? (`No`)

## Risks and Mitigations

- Risk: Tools that previously accepted `skillIdsJson` from LLM now ignore it and use `skillScope` only.
  - Mitigation: `skillScope` is set from the same source (`state.selectedSkillIds`) — no behavior change for correctly configured sessions.
- Risk: Generic skill model building now calls `tryBuildGenericModelWithLlm` which requires LLM access.
  - Mitigation: `createChatModel(0)` returns null if no API key configured; the tool returns a clear error message.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: backend/src/agent-langgraph/tools.ts [hunk 1]
```diff
@@ -219,31 +222,29 @@ export function createExtractDraftParamsTool(skillRuntime: AgentSkillRuntime) {
...
+      const skillIds = configurable.skillScope;
 
-      const model = await skillRuntime.buildModel(draftState, skillIds);
+      const model = await skillRuntime.buildModel(draftState, skillIds, {
+        message: state?.lastUserMessage || '',
+        locale: state?.locale || 'zh',
+      });
       if (!model) {
         throw new Error('Model build returned undefined — draft may be incomplete. Try running extract_draft_params again with more explicit parameters.');
       }
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-langgraph/tools.ts:250]: Fixed in 2d8ddef. buildModel now throws specific bilingual error when LLM not configured.

File: backend/src/agent-runtime/index.ts [hunk 2]
```diff
@@ -495,12 +495,33 @@ export class AgentSkillRuntime {
...
+    }
+
+    // Generic skill: fall back to LLM model builder
+    if (plugin.id === 'generic' && context) {
+      const { tryBuildGenericModelWithLlm } = await import('../agent-skills/structure-type/generic/llm-model-builder.js');
+      const { createChatModel } = await import('../utils/llm.js');
+      const llm = createChatModel(0);
+      if (!llm) {
+        return undefined;
+      }
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-runtime/index.ts:522]: Fixed in 2d8ddef — buildModel() now throws a specific bilingual error when createChatModel returns null, instead of returning undefined silently. The build_model tool propagates this error directly so users see the real reason.

File: frontend/src/components/chat/ai-console.tsx [hunk 3]
```diff
@@ -2018,10 +2018,11 @@ export function AIConsole() {
           return
         }
         const payload = await response.json()
-        if (!active || !Array.isArray(payload)) {
+        const skillsArray = Array.isArray(payload) ? payload : Array.isArray(payload?.skills) ? payload.skills : null
+        if (!active || !skillsArray) {
           return
         }
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply frontend/src/components/chat/ai-console.tsx:2028]: Valid point. Fixed in 2d8ddef — when skillsArray is null (unexpected payload shape), we now call setSkillsLoaded(true) before returning so the UI does not hang in a loading state.

File: frontend/src/components/chat/capability-settings-panel.tsx [hunk 4]
```diff
@@ -160,8 +160,9 @@ export function CapabilitySettingsPanel() {
         return
       }
       const payload = await response.json()
-      if (active && Array.isArray(payload)) {
-        setAvailableSkills(payload as AgentSkillSummary[])
+      const skillsArray = Array.isArray(payload) ? payload : Array.isArray(payload?.skills) ? payload.skills : null
+      if (active && skillsArray) {
+        setAvailableSkills(skillsArray as AgentSkillSummary[])
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply frontend/src/components/chat/capability-settings-panel.tsx]: Same fix applied in 2d8ddef — setSkillsLoaded(true) is now called even when the payload has an unexpected shape, preventing the settings panel from hanging.

File: backend/src/agent-runtime/index.ts [hunk 5]
```diff
@@ -495,12 +495,33 @@ export class AgentSkillRuntime {
...
+      return tryBuildGenericModelWithLlm(
+        llm, context.message || '', state, context.locale || 'zh',
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-runtime/index.ts:524]: Fixed in 2d8ddef — same as the Copilot review: buildModel() now throws a specific error when LLM is not configured for the generic skill fallback.

File: backend/src/agent-langgraph/streaming.ts [hunk 6]
```diff
@@ -73,6 +73,20 @@ function extractTextContent(content: unknown): string {
...
+ * { skillId: "..." } in their output JSON.
+ */
+function extractSkillIdFromContent(content: string): string | undefined {
+  try {
+    const parsed = JSON.parse(content);
+    return typeof parsed?.skillId === 'string' ? parsed.skillId : undefined;
+  } catch {
+    return undefined;
+  }
+}
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-langgraph/streaming.ts:88]: The try-catch approach is intentional and sufficient here. Tool output content is always a JSON string controlled by our own tool implementations (detect_structure_type, run_analysis, etc.), so JSON.parse will always succeed on valid tool output. When it fails (e.g., error messages or truncated content), the catch returns undefined gracefully. Adding a startsWith check is redundant — if the content does not start with '{', JSON.parse would fail anyway and we'd get undefined. The performance concern is not relevant: tool outputs are small JSON objects (< 1KB), not large strings.

**Review by guyi2000 — APPROVED** [review]
